spec.md

```diff
@@ -2354,6 +2364,15 @@ message NodeExpandVolumeRequest {
   // plugin MAY expand the volume to its maximum capacity.
   // This field is OPTIONAL.
   CapacityRange capacity_range = 3;
+
+  // Volume capability describing how the CO intends to use this volume.
+  // This allows SP to determine if volume is being used as a block
+  // device or mounted file system. For example - if volume is being
+  // used as a block device the SP MAY choose to skip expanding the
```

Comment:
It looks like whether the `NodeExpand` can do work is psuedo-determined in two places
1) here, based on the volume capability
2) setting `node_expansion_required` on the response to `ControllerExpand`

Can you explain why you might want to set one or not the other, or when they would differ? Basically: why do we need both, can we simplify the spec to use only one of them?

Reply:
The call to `NodeExpand` is still made if `node_expansion_required` is set to true. It is upto SP to process a block device node expansion and determine how expansion should work. For example:

1. In case of EBS or GCE-PD raw block devices, `NodeExpand` call might be a NO-OP and as such may never be called.
2. For something like iSCSI raw block devices though, even if there is no file system on the device it may still require rescan of the LUN, so as OS can see updated size.

Reply:
Since it is the `SP` that processes `ControllerExpand` and has `Block/FS` information at that level and sets the `node_expansion_required`. In case 1) couldn't those drivers set `node_expansion_required=false` and in case 2) those drivers set `node_expansion_required=true`. When `node_expansion_required=false` the `NodeExpand` should never get called right?

Then the `NodeExpand` doesn't really need the capability to decide anything anymore. I feel like maybe I'm misunderstanding something here. Could you explain why `node_expansion_required` is not enough of a signal for doing a `NodeExpand`? Is there the possibility of `node_expansion_required=true` and there will be different behavior for either `Block or FS`

Reply:
> Is there the possibility of node_expansion_required=true and there will be different behavior for either Block or FS

Yes there is! For example - an iSCSI implementation may return `node_expansion_required` as `true` in `ControllerExpandVolume` call for raw block devices but actual implementation of `NodeExpandVolume` call for iSCSI might be different between whether device is raw block or FS. For example - if device is raw block, it could just do a rescan of the LUN. But device has FS on it, it would rescan the LUN and also expand the filesystem.

We must also consider case of volume types that don't necessarily implement a control-plane volume expansion (such as ceph-csi). In that case, actual implementation of `NodeExpandVolume` might be different based on block or FS.

Reply:
ah I see, this answers my question. Thanks!